File: swiftmap/mapops/legend.py
"""
The legend overlay's Python half: display options and manual overrides written
into the legend_config trait. Derivation and rendering live in src/legend.js.
"""
from typing import Any, Dict, List, Optional

from .._warnings import warn

_LEGEND_POSITIONS = frozenset({
    "top-left", "top-center", "top-right", "left-center", "right-center",
    "bottom-left", "bottom-center", "bottom-right"})


def configure_legend(self, *, title: Optional[str] = None,
                     position: Optional[str] = None,
                     auto: Optional[bool] = None,
                     scope: Optional[str] = None,
                     dim_hidden: Optional[bool] = None,
                     show: Optional[bool] = None) -> "Map":
    """
    Configures the legend overlay. Only the options given change.

    Parameters
    ----------
    title : str, optional
        Heading text. Default "Legend".
    position : str, optional
        One of the eight anchors ('bottom-left' default): 'top-left', 'top-center',
        'top-right', 'left-center', 'right-center', 'bottom-left', 'bottom-center',
        'bottom-right'.
    auto : bool, optional
        False turns derivation off entirely: the legend is exactly what
        `legend_add` added. Default True.
    scope : {'all', 'visible'}, optional
        'all' (default) lists every layer that could be shown, dimming hidden
        ones -- a legend is the map's vocabulary. 'visible' tracks the screen:
        unchecking a layer drops its row.
    dim_hidden : bool, optional
        Under scope 'all', whether hidden layers render dimmed. Default True;
        False gives the print-style static legend.
    show : bool, optional
        Convenience for setting `show_legend`.

    Examples
    --------
    >>> m.configure_legend(show=True, position="bottom-right", title="Key")
    >>> m.configure_legend(auto=False)     # manual entries only
    """
    cfg = dict(self.legend_config)
    if title is not None:
        cfg["title"] = title
    if position is not None:
        if position not in self._LEGEND_POSITIONS:
            warn(f"configure_legend: unknown position {position!r}; expected one "
                 f"of {', '.join(sorted(self._LEGEND_POSITIONS))}. Ignored.")
        else:
            cfg["position"] = position
    if auto is not None:
        cfg["auto"] = bool(auto)
    if scope is not None:
        if scope not in ("all", "visible"):
            warn(f"configure_legend: scope must be 'all' or 'visible', got "
                 f"{scope!r}. Ignored.")
        else:
            cfg["scope"] = scope
    if dim_hidden is not None:
        cfg["dim_hidden"] = bool(dim_hidden)
    self.legend_config = cfg
    if show is not None:
        self.show_legend = bool(show)
    return self


def legend_add(self, label: str, *, group: Optional[str] = None,
               shape: Optional[str] = None, color: Optional[str] = None,
               fill_color: Optional[str] = None,
               colormap: Optional[str] = None,
               vmin: Optional[float] = None, vmax: Optional[float] = None,
               color_bins: Optional[List[float]] = None,
               categories: Any = None,
               layer: Optional[str] = None) -> "Map":
    """
    Adds a manual legend entry -- the freedom hatch when derivation misses.

    Manual entries are the user's own claims: scope never drops them and auto=False
    keeps them, so `configure_legend(auto=False)` plus `legend_add` calls builds a
    fully hand-authored legend. Everything is resolved here into frontend-ready
    data, so a manual entry renders identically to a derived one.

    Parameters
    ----------
    label : str
        The entry's text.
    group : str, optional
        Section heading to file it under; ungrouped entries lead the legend.
    shape : {'circle', 'pin', 'line', 'polygon', 'square'}, optional
        Swatch glyph. Default 'square' for a plain colour entry.
    color, fill_color : str, optional
        Swatch stroke and fill (fill defaults to the stroke).
    colormap : str, optional
        Makes this a ramp entry (with vmin/vmax, default 0..1), or with
        `color_bins` a discrete classed entry.
    vmin, vmax : float, optional
        Ramp endpoint labels.
    color_bins : list of float, optional
        Bin edges for a classed entry; colours sample the colormap.
    categories : dict or list, optional
        Makes this a category list: {value: colour} taken as given, or a list of
        values coloured from the palette exactly as `color_col` would.
    layer : str, optional
        A layer id or name to bind to: the entry dims (or drops, under scope
        'visible') with that layer, for when a manual row relabels a live one.

    Examples
    --------
    >>> m.legend_add("Restricted zone", shape="polygon", color="#f00",
    ...              fill_color="#ff000044", group="Zones")
    >>> m.legend_add("Threat score", colormap="turbo", vmin=0, vmax=100)
    >>> m.legend_add("Sightings", categories=["confirmed", "probable"])
    """
    from .._colormaps import (COLORMAPS, DEFAULT_COLORMAP, _category_table,
                              _label_num, rgb_hex, bins_block)
    entry: Dict[str, Any] = {"label": str(label)}
    if group is not None:
        entry["group"] = str(group)
    if layer is not None:
        entry["layer"] = layer

    if categories is not None:
        if isinstance(categories, dict):
            items = [{"value": str(k), "color": v} for k, v in categories.items()]
        else:
            values = [str(v) for v in categories]
            table = _category_table(len(values), colormap, quiet=False)
            items = [{"value": v, "color": rgb_hex(table[i])}
                     for i, v in enumerate(values)]
        entry.update({"kind": "categories", "items": items})
    elif colormap is not None and color_bins is not None:
        entry.update(bins_block(colormap, color_bins))
    elif colormap is not None:
        anchors = COLORMAPS.get(str(colormap).lower())
        if anchors is None:
            warn(f"legend_add: unknown colormap {colormap!r}; using "
                 f"{DEFAULT_COLORMAP!r}.")
            anchors = COLORMAPS[DEFAULT_COLORMAP]
        entry.update({"kind": "ramp", "anchors": list(anchors),
                      "vmin": _label_num(vmin if vmin is not None else 0),
                      "vmax": _label_num(vmax if vmax is not None else 1)})
    else:
        entry.update({"kind": "swatch", "shape": shape or "square",
                      "color": color or "#3388ff",
                      "fillColor": fill_color or color or "#3388ff"})

    cfg = dict(self.legend_config)
    cfg["add"] = list(cfg.get("add", [])) + [entry]
    self.legend_config = cfg
    return self


def legend_remove(self, target: Optional[str] = None, *,
                  group: Optional[str] = None,
                  id: Optional[str] = None) -> "Map":
    """
    Suppresses legend entries matching the criteria -- derived or manual.

    A persistent matcher, not a one-shot deletion: it keeps suppressing across
    every re-derivation, so a hidden row stays hidden however the layers change.
    Any combination of label, group and id must all match.

    Examples
    --------
    >>> m.legend_remove("Sites")            # by label
    >>> m.legend_remove(group="Debug")      # a whole section
    """
    if target is None and group is None and id is None:
        warn("legend_remove: nothing to match -- give a label, group=, or id=.")
        return self
    matcher: Dict[str, Any] = {}
    if target is not None:
        matcher["label"] = str(target)
    if group is not None:
        matcher["group"] = str(group)
    if id is not None:
        matcher["id"] = id
    cfg = dict(self.legend_config)
    cfg["remove"] = list(cfg.get("remove", [])) + [matcher]
    self.legend_config = cfg
    return self


def legend_clear(self) -> "Map":
    """Drops every manual entry and suppression; display options are kept."""
    cfg = {k: v for k, v in self.legend_config.items()
           if k not in ("add", "remove")}
    self.legend_config = cfg
    return self


def legend_html(self) -> str:
    """
    Returns a plain HTML listing of layer names with colour chips, or "" when
    `show_legend` is False.

    Legacy beside the real legend: the map now draws a full overlay itself --
    derived entries, ramps and size keys from color_col/radius_col, and the
    `legend_add`/`legend_remove` overrides -- switched on with `show_legend=True`
    and configured through `configure_legend`. This property predates all of that
    and renders none of it; it survives only for layouts that embed a simple
    list outside the map. Expect it to be retired.
    """
    if not self.show_legend:
        return ""
    items = []
    for l in self.layers:
        name = l.get("name", "Layer")
        color = l.get("color", "#3388ff")
        items.append(f"<div><span style='background:{color};width:10px;height:10px;display:inline-block;margin-right:5px;'></span>{name}</div>")
    return "".join(items)

File: src/legend.js
// The legend: derived from the same layer state everything else renders from, with
// declarative overrides on top. Deliberately model-free -- pure data in, DOM out --
// so a plain-JS consumer of dist/index.js gets the whole feature, and the anywidget
// glue in map.js is a few lines. (sidebar.js still takes `model` and is filed for
// extraction; this module must never need that unpicking.)
//
// The pipeline: deriveLegendSpec(layers, groupConfigs, config) walks the layers into
// entries (skipped entirely when config.auto === false), applies the persistent
// remove-matchers, appends the manual adds, and returns a spec that renderLegend
// turns into DOM. Nothing here knows about colormaps: ramp/category/bin entries
// arrive with their colours already resolved (Python resolves at the add_* boundary,
// manual entries at legend_add), so there is no anchor table to drift.

import { isLayerEffectiveVisible } from "./patch.js";

const GLYPHS = {
    circle_markers: "circle",
    markers: "pin",
    polyline: "line",
    polygon: "polygon",
    circle: "circle",
};

function swatchEntry(layer, hidden) {
    return {
        kind: "swatch",
        label: layer.name || "Layer",
        shape: GLYPHS[layer.type] || "square",
        color: layer.color || "#3388ff",
        fillColor: layer.fillColor || layer.fill_color || layer.color || "#3388ff",
        hidden,
    };
}

// A data-driven block recorded at add time ({kind, anchors|items|edges+colors, ...})
// becomes the layer's entry as-is; the layer only contributes label and visibility.
function blockEntry(layer, hidden) {
    return { ...layer.legend, label: layer.name || "Layer", hidden };
}

function entriesForLayer(layer, groupConfigs) {
    if (layer.type === "basemap") return [];
    const hidden = !isLayerEffectiveVisible(layer, groupConfigs);
    if (layer.type === "group") {
        // A collection: one entry per geometry part, same label by design -- the
        // glyphs are what tell them apart, matching how the parts render.
        return (layer.layers || [])
            .filter(sub => GLYPHS[sub.type])
            .map(sub => sub.legend
                ? blockEntry({ ...sub, name: layer.name }, hidden)
                : swatchEntry({ ...sub, name: layer.name }, hidden));
    }
    if (!GLYPHS[layer.type]) return [];
    const entries = [layer.legend ? blockEntry(layer, hidden) : swatchEntry(layer, hidden)];
    // radius_col records a size key beside the colour story: both encodings on the
    // map deserve both explanations in the legend.
    if (layer.legend_size) {
        entries.push({ ...layer.legend_size,
                       label: layer.legend_size.field || layer.name || "Size", hidden });
    }
    return entries;
}

// Identical data-driven payloads collapse into one row. Grouping points by a column
// gives every sub-layer the same ramp; a ramp per sub-layer is noise, and the field
// name is the honest label for the shared mapping. The survivor keeps the first
// occurrence's position and hides only when every contributor is hidden.
function payloadKey(entry) {
    // Identity fields stay out of the key: the whole point is that entries from
    // DIFFERENT layers collapse when their mapping payload is the same.
    const { label, hidden, layerId, layer, group, ...payload } = entry;
    return JSON.stringify(payload);
}

function dedupeDataEntries(groups) {
    const seen = new Map();   // payload key -> surviving entry
    for (const group of groups) {
        group.entries = group.entries.filter(entry => {
            if (entry.kind === "swatch") return true;
            const key = payloadKey(entry);
            const survivor = seen.get(key);
            if (!survivor) {
                seen.set(key, entry);
                if (entry.field) entry.label = entry.field;
                return true;
            }
            survivor.hidden = survivor.hidden && entry.hidden;
            return false;
        });
    }
    return groups;
}

function matcherHits(matcher, entry, groupName) {
    if (!matcher) return false;
    let constrained = false;
    if (matcher.label != null) {
        constrained = true;
        if (entry.label !== matcher.label) return false;
    }
    if (matcher.group != null) {
        constrained = true;
        if (groupName !== matcher.group) return false;
    }
    if (matcher.id != null) {
        constrained = true;
        if (entry.layerId !== matcher.id) return false;
    }
    return constrained;
}

// A ramp endpoint as it should read in a legend: 10, not 10.000000001, and
// 0.003757, not 0.0037567270919680595. Four significant figures, integers whole,
// anything non-numeric passed through. Python rounds its side before composing
// a block (_label_num), but legends derived or hand-built in JS arrive raw.
export function formatBound(value) {
    const n = Number(value);
    if (value == null || value === "" || !isFinite(n)) return String(value);
    return Number.isInteger(n) ? String(n) : String(Number(n.toPrecision(4)));
}

export function deriveLegendSpec(layers, groupConfigs, config) {
    const cfg = config || {};
    const groups = [];
    const byName = new Map();
    const groupFor = name => {
        if (!byName.has(name)) {
            const group = { name, entries: [] };
            byName.set(name, group);
            groups.push(group);
        }
        return byName.get(name);
    };

    if (cfg.auto !== false) {
        for (const layer of layers || []) {
            for (const entry of entriesForLayer(layer, groupConfigs || {})) {
                entry.layerId = layer.id;
                if (cfg.scope === "visible" && entry.hidden) continue;
                groupFor(layer.layer_group || "Layers").entries.push(entry);
            }
        }
        dedupeDataEntries(groups);
    }

    // Persistent suppression: matchers outlive every re-derivation, which is the
    // difference from a registry remove that the next add would just repopulate.
    const removes = cfg.remove || [];
    if (removes.length > 0) {
        for (const group of groups) {
            group.entries = group.entries.filter(
                entry => !removes.some(m => matcherHits(m, entry, group.name)));
        }
    }

    // Manual entries: the user's own claims. scope never drops them; a `layer`
    // binding makes one follow a live layer's visibility (and vanish with it under
    // scope "visible"), for when a manual row is really a relabelling.
    for (const added of cfg.add || []) {
        const entry = { hidden: false, ...added };
        if (entry.layer != null) {
            const bound = (layers || []).find(
                l => l.id === entry.layer || l.name === entry.layer);
            entry.hidden = !bound || !isLayerEffectiveVisible(bound, groupConfigs || {});
            if (cfg.scope === "visible" && entry.hidden) continue;
        }
        if (removes.some(m => matcherHits(m, entry, entry.group || ""))) continue;
        groupFor(entry.group || "").entries.push(entry);
    }

    const populated = groups.filter(g => g.entries.length > 0);
    return { title: cfg.title || "Legend", groups: populated };
}

// --- rendering ------------------------------------------------------------------------
// DOM built with createElement/textContent throughout: labels and category values come
// from user data and must never be parsed as HTML.

function div(styles, text) {
    const el = document.createElement("div");
    Object.assign(el.style, styles);
    if (text != null) el.textContent = text;
    return el;
}

function glyph(entry) {
    if (entry.shape === "line") {
        return div({ width: "20px", height: "4px", background: entry.color,
                     marginRight: "6px", flex: "none" });
    }
    if (entry.shape === "pin") {
        const el = document.createElement("span");
        el.style.marginRight = "6px";
        el.style.flex = "none";
        const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
        svg.setAttribute("width", "12");
        svg.setAttribute("height", "14");
        svg.setAttribute("viewBox", "0 0 24 28");
        const path = document.createElementNS("http://www.w3.org/2000/svg", "path");
        path.setAttribute("d",
            "M12 0C5.4 0 0 5.4 0 12c0 9 12 16 12 16s12-7 12-16C24 5.4 18.6 0 12 0z");
        path.setAttribute("fill", entry.color);
        svg.appendChild(path);
        el.appendChild(svg);
        return el;
    }
    // circle / polygon / square: fill inside a border, which is how areas draw.
    const radius = entry.shape === "circle" ? "50%"
        : entry.shape === "polygon" ? "2px" : "0";
    return div({ width: "12px", height: "12px", background: entry.fillColor,
                 border: `2px solid ${entry.color}`, borderRadius: radius,
                 marginRight: "6px", flex: "none", boxSizing: "border-box" });
}

function rampRow(entry) {
    const row = div({ marginTop: "5px" });
    row.appendChild(div({}, entry.label));
    const stops = (entry.anchors || []).map((color, i, all) =>
        `${color} ${all.length > 1 ? (i / (all.length - 1)) * 100 : 0}%`);
    row.appendChild(div({
        width: "120px", height: "12px", borderRadius: "2px",
        backgroundImage: `linear-gradient(to right, ${stops.join(", ")})`,
    }));
    const ends = div({ display: "flex", justifyContent: "space-between", width: "120px",
                       fontSize: "11px", color: "#555" });
    ends.appendChild(div({}, formatBound(entry.vmin)));
    ends.appendChild(div({}, formatBound(entry.vmax)));
    row.appendChild(ends);
    return row;
}

const MAX_CATEGORY_ROWS = 12;

function categoriesRow(entry) {
    const row = div({ marginTop: "5px" });
    row.appendChild(div({}, entry.label));
    const items = entry.items || [];
    for (const item of items.slice(0, MAX_CATEGORY_ROWS)) {
        const line = div({ display: "flex", alignItems: "center", marginTop: "3px",
                           marginLeft: "8px" });
        line.appendChild(glyph({ shape: "square", color: item.color, fillColor: item.color }));
        line.appendChild(div({}, String(item.value)));
        row.appendChild(line);
    }
    if (items.length > MAX_CATEGORY_ROWS) {
        row.appendChild(div({ marginLeft: "8px", marginTop: "3px", color: "#555" },
            `+ ${items.length - MAX_CATEGORY_ROWS} more`));
    }
    return row;
}

function binsRow(entry) {
    const row = div({ marginTop: "5px" });
    row.appendChild(div({}, entry.label));
    const edges = entry.edges || [];
    const colors = entry.colors || [];
    const labelFor = i => i === 0 ? `< ${edges[0]}`
        : i === edges.length ? `≥ ${edges[edges.length - 1]}`
        : `${edges[i - 1]} – ${edges[i]}`;
    colors.forEach((color, i) => {
        const line = div({ display: "flex", alignItems: "center", marginTop: "3px",
                           marginLeft: "8px" });
        line.appendChild(glyph({ shape: "square", color, fillColor: color }));
        line.appendChild(div({}, labelFor(i)));
        row.appendChild(line);
    });
    return row;
}

// A size key is a statement, not a drawing: "● size ∝ field (min – max)". The glyph
// is fixed and nothing in the row derives from radius_range or the data's spread --
// legend CSS pixels are not map pixels at any zoom, so drawn sample circles would
// assert a precision that does not exist. The row names the encoding and its domain.
function sizesRow(entry) {
    const row = div({ display: "flex", alignItems: "center", marginTop: "5px" });
    row.appendChild(div({ marginRight: "6px", flex: "none", color: "#666" }, "●"));
    const range = entry.vmin != null && entry.vmax != null
        ? ` (${formatBound(entry.vmin)} – ${formatBound(entry.vmax)})` : "";
    row.appendChild(div({}, `size ∝ ${entry.field || entry.label}${range}`));
    return row;
}

function swatchRow(entry) {
    const row = div({ display: "flex", alignItems: "center", marginTop: "5px" });
    row.appendChild(glyph(entry));
    row.appendChild(div({}, entry.label));
    return row;
}

// Collapse state, per container rather than module scope: the sidebar keys its
// collapsedPaths at module level and two maps on one page share it -- a filed bug
// this deliberately does not inherit. Keyed by group name, surviving the full
// re-render every sync performs.
const collapsedByContainer = new WeakMap();

export function renderLegend(container, spec, options = {}) {
    container.innerHTML = "";
    const dimHidden = options.dimHidden !== false;
    let collapsed = collapsedByContainer.get(container);
    if (!collapsed) {
        collapsed = new Set();
        collapsedByContainer.set(container, collapsed);
    }
    container.appendChild(div({
        fontSize: "13px", fontWeight: "bold", borderBottom: "2px solid #eee",
        paddingBottom: "4px", marginBottom: "4px",
    }, spec.title));

    for (const group of spec.groups) {
        const isCollapsed = group.name && collapsed.has(group.name);
        if (group.name) {
            // The sidebar's affordance exactly: an arrow that folds the section.
            const header = div({ fontWeight: "bold", marginTop: "6px",
                                 cursor: "pointer", userSelect: "none" });
            header.textContent = `${isCollapsed ? "▸" : "▾"} ${group.name}`;
            header.addEventListener("click", () => {
                if (collapsed.has(group.name)) collapsed.delete(group.name);
                else collapsed.add(group.name);
                renderLegend(container, spec, options);
            });
            container.appendChild(header);
        }
        if (isCollapsed) continue;
        for (const entry of group.entries) {
            const row = entry.kind === "ramp" ? rampRow(entry)
                : entry.kind === "categories" ? categoriesRow(entry)
                : entry.kind === "bins" ? binsRow(entry)
                : entry.kind === "sizes" ? sizesRow(entry)
                : swatchRow(entry);
            // Dimmed, not dropped: under scope "all" the legend is the map's whole
            // vocabulary, and the dim is what still tells the current screen state.
            if (entry.hidden && dimHidden) row.style.opacity = "0.5";
            container.appendChild(row);
        }
    }
    return container;
}
